scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -201,108 +260,275 @@ def _highly_variable_genes_single_batch(
     A DataFrame that contains the columns
     `highly_variable`, `means`, `dispersions`, and `dispersions_norm`.
     """
-    X = adata.layers[layer] if layer is not None else adata.X
+    data = _get_obs_rep(adata, layer=layer)
     if flavor == "seurat":
-        X = X.copy()
+        data = data.copy()
         if "log1p" in adata.uns_keys() and adata.uns["log1p"].get("base") is not None:
-            X *= np.log(adata.uns["log1p"]["base"])
-        # use out if possible. only possible since we copy X
-        if isinstance(X, np.ndarray):
-            np.expm1(X, out=X)
+            data *= np.log(adata.uns["log1p"]["base"])
+        # use out if possible. only possible since we copy the data matrix
+        if isinstance(data, np.ndarray):
+            np.expm1(data, out=data)
         else:
-            X = np.expm1(X)
+            data = np.expm1(data)
 
-    mean, var = materialize_as_ndarray(_get_mean_var(X))
+    mean, var = _get_mean_var(data)
     # now actually compute the dispersion
     mean[mean == 0] = 1e-12  # set entries equal to zero to small value
     dispersion = var / mean
     if flavor == "seurat":  # logarithmized mean as in Seurat
         dispersion[dispersion == 0] = np.nan
         dispersion = np.log(dispersion)
         mean = np.log1p(mean)
+
     # all of the following quantities are "per-gene" here
-    df = pd.DataFrame()
-    df["means"] = mean
-    df["dispersions"] = dispersion
+    df: pd.DataFrame | DaskDataFrame
+    if isinstance(data, DaskArray):
+        import dask.array as da
+        import dask.dataframe as dd
+
+        df = dd.from_dask_array(
+            da.vstack((mean, dispersion)).T, columns=["means", "dispersions"]
+        )
```

Comment:
```suggestion
    if isinstance(data, DaskArray):
        import dask

        df = pd.DataFrame(dict(zip(["means", "dispersions"], dask.compute(mean, dispersion))))
```

Re-my previous thought "I think this will be faster if we use pandas instead of dask.dataframe", with this change I see about a 4x speed up. Note that presumably the most computationally expensive part of this whole thing is the mean and dispersion calculation, which happens the same in both cases.

Here is the script I used to test: https://gist.github.com/ivirshup/ecf131b6ba8b094c04bd4a275ad7a804

And here is the branch with my code: https://github.com/scverse/scanpy/tree/dask-hvg-w-pandas

I haven't dug too deeply in yet so:

* I don't know the effect of using batches
* I've only looked at Seurat
* I haven't looked at the effect of using different chunk sizes

One red flag I noticed while playing around with this was that the dask dataframe returned with `inplace=False` had ~1000 high level graph nodes when I used a chunk size of 1000 cells. That seems like it could be incurring a lot of overhead. In addition, the majority of time to calculate actually came before this dask dataframe was returned, due to an intermediate computation.

Surprisingly, I was seeing increasing times to compute as I increased chunk size – but haven't run this enough to be that confident in this result.